server/src/main/java/org/elasticsearch/search/profile/SearchProfileResultsBuilder.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.search.profile;
+
+import org.elasticsearch.search.SearchPhaseResult;
+import org.elasticsearch.search.fetch.FetchSearchResult;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Profile results for the query phase run on all shards.
+ */
+public class SearchProfileResultsBuilder {
+    public static SearchProfileResultsBuilder build(Map<String, SearchProfileQueryPhaseResult> queryPhaseResults) {
```

Comment:
I am a little confused about this part -- probably related to naming. Could we just have a constructor `new SearchProfileResultsBuilder(Map<String, SearchProfileQueryPhaseResult>)` instead of this static method? I'm not sure why we need the special `NOT_PROFILING` object instead of just passing `null` in the original place in `SearchPhaseController`.

Reply:
I was trying to get the null check out of search phase controller so it was easy to unit test without all of SearchPhaseController. This way there isn't a null check. I liked that the old code didn't allocate anything if we didn't profile and wanted to keep that too, but have it here.

Another idea is to make the `build` method static and have it take a nullable builder param. Or just revert the whole change and keep the null check in SearchPhaseController. I just, like, didn't like it there. I'm happy anytime I can get things *out* of SearchPhaseController.

Reply:
Since this change has so many "moving parts" already, I'd be in favor of keeping the old strategy to keep it well-scoped. I also find it easier to understand. If we want we could follow-up with a refactor to simplify `SearchPhaseController` a bit.

Reply:
Yeah. I think it was clearer that way but I'm not the only one that has to read it. I've changed it the way you asked.